### ISSUE TITLE
Keeps one CPU thread busy

### ISSUE THREAD

**Issue by raducrisan1** [icsharpcode/RefactoringEssentials#215]
Hi,

Having a pretty large code base I observed that a CPU thread is always used. I waited for more than 10 minutes to see if the behavior disappears, but it didn't happen. Eventually I disabled the extension and opened again the solution. After 1 or 2 minutes I didn't observe any CPU usage. Host machine: Windows 8.1 with VS 2015 Update 2.

**Comment by ddobrev** [comment]
This might be related to an issue I have. I work on https://github.com/mono/CppSharp which is a fairly large solution of 51 projects. When I make a single change and save the file, my CPU usage jumps to 50 % for at least 15 seconds. It may not seem as much but during regular work changes are saved much faster and I have seen the difference stack. There's definitely something wrong because while I understand the AST must be updated when a file is changed, this jump in CPU usage occurs even if I change the C++ part of my project which has nothing to do with the area Refactoring Essentials covers.
Note: in case you want to test with CppSharp, please have in mind that it does not contain a static solution file, it must be generated instead. Instructions to do so you can find at https://github.com/mono/CppSharp/blob/master/docs/GettingStarted.md . In case you find them confusing, all you need to do is go to build/ and run DownloadDeps.bat, then GenerateProjects.bat with selection VS 2015.

**Comment by ddobrev** [comment]
Any comment? Is this perhaps related to https://github.com/icsharpcode/RefactoringEssentials/issues/159 ?

**Comment by Rpinski** [comment]
Indeed, after opening a big solution it takes several minutes until analysis has finished. We plan to configure the analyzers to asynchronous execution (a new feature in Roslyn 1.2) and see if that decreases the time amount needed to finish the solution.

About analysis when editing C++ code: All analyzers are triggered by Roslyn and we have actually no control over the moment it happens. It will need some tests to see if the problem is on our or on Roslyn's side.

**Comment by ddobrev** [comment]
Thank you for taking measures. In case you decide to test with CppSharp and have problems generating the solution, I'll be glad to help.

**Comment by ddobrev** [comment]
I would also like to get some more insight into the way Refactoring Essentials works. To begin with, what and why does it need to analyse given that Roslyn has already constructed the AST of the project and, I assume, provides it to add-in developers?

**Comment by Rpinski** [comment]
"Analyze" means analyzing the AST generated by Roslyn. This is what a Roslyn-based diagnostic analyzer usually does.

**Comment by ddobrev** [comment]
I understand, thank you. I hope to see the optimised analysers soon.

**Comment by ddobrev** [comment]
My VS regularly stops responding while loading large solutions with RE on.Could you please provide an estimate when these optimised analysers will be complete?

**Comment by Rpinski** [comment]
Concurrent execution is already activated for all RE analyzers in development builds (see c49e79795f120654b62e342b875e1c414747cbd3) and this will be part of release 4.1, which we are about to finish today or tomorrow. Unfortunately I don't see the high CPU load going away with it completely. But please test whether the time with high CPU load will decrease, at least.
